src/Time/TimeSteppers/RungeKutta4.hpp

```diff
@@ -0,0 +1,187 @@
+                const TimeDelta& time_step) const noexcept;
+
+  template <typename Vars, typename DerivVars>
+  void dense_update_u(gsl::not_null<Vars*> u,
+                      const History<Vars, DerivVars>& history,
+                      double time) const noexcept;
+
+  uint64_t number_of_substeps() const noexcept override;
+
+  size_t number_of_past_steps() const noexcept override;
+
+  double stable_step() const noexcept override;
+
+  TimeStepId next_time_id(const TimeStepId& current_id,
+                          const TimeDelta& time_step) const noexcept override;
+
+  WRAPPED_PUPable_decl_template(RungeKutta4);  // NOLINT
+
+  explicit RungeKutta4(CkMigrateMessage* /*unused*/) noexcept {}
+
+  // clang-tidy: do not pass by non-const reference
+  void pup(PUP::er& p) noexcept override {  // NOLINT
+    TimeStepper::Inherit::pup(p);
+  }
+};
+
+inline bool constexpr operator==(const RungeKutta4& /*lhs*/,
+                                 const RungeKutta4& /*rhs*/) noexcept {
+  return true;
+}
+
+inline bool constexpr operator!=(const RungeKutta4& /*lhs*/,
+                                 const RungeKutta4& /*rhs*/) noexcept {
+  return false;
+}
+
+template <typename Vars, typename DerivVars>
+void RungeKutta4::update_u(
+    const gsl::not_null<Vars*> u,
+    const gsl::not_null<History<Vars, DerivVars>*> history,
+    const TimeDelta& time_step) const noexcept {
+  const size_t substep = history->size() - 1;
+  const auto& vars = (history->end() - 1).value();
+  const auto& dt_vars = (history->end() - 1).derivative();
+  const auto& U0 = history->begin().value();
+
+  switch (substep) {
+    case 0: {
+      // from (17.1.3) of Numerical Recipes 3rd Edition
+      // v^(1) = u^n + dt*RHS(u^n,t^n)/2
+      // On entry V = u^n, U0 = u^n, rhs0 = RHS(u^n, t^n),
+      // time = t^n
+      *u += (time_step.value() * dt_vars) / 2;
+      // On exit v = v^(1), time = t^n + (1/2)*dt
+      break;
+    }
+    case 1: {
+      // from (17.1.3) of Numerical Recipes 3rd Edition
+      // v^(2) = u^n + dt * RHS(v^(1), t^n + (1/2)*dt)/2
+      // On entry V = v^(1), U0 = u^n, rhs0 = RHS(v^(1), t^n + dt/2),
+      // time = t^n + dt
+      *u = U0 + (time_step.value() * dt_vars) / 2;
+      // On exit v = v^(2), time = t^n + (1/2)*dt
+      break;
+    }
+    case 2: {
+      // from (17.1.3) of Numerical Recipes 3rd Edition
+      // v^(3) = u^n + dt*RHS(v^(2), t^n + (1/2)*dt))
+      // On entry V = v^(2), U0 = u^n, rhs0 = RHS(v^(2), t^n + (1/2)*dt),
+      // time = t^n + (1/2)*dt
+      *u = U0 + time_step.value() * dt_vars;
+      // On exit v = v^(3), time = t^n + dt
+      break;
+    }
+    case 3: {
+      // from (17.1.3) of Numerical Recipes 3rd Edition
+      // v^(4) = u^n + dt*RHS(v^(3), t+dt)
+      // On entry V = v^(3), U0 = u^n, rhs0 = RHS(v^(3), t^n + dt),
+      // time = t^n + (1/2)*dt
+      *u = U0 + time_step.value() * dt_vars;
+      // On exit v = v^(4), time = t^n + dt
```

Comment:
Can update `*u` directly here to get the final step by moving the combining code below into the `case 3` section.